scipy/integrate/_tanhsinh.py

```diff
@@ -760,3 +760,308 @@ def _tanhsinh_iv(f, a, b, log, maxfun, maxlevel, minlevel,
+        S[i], E[i] = tmp[:-1]
+        nfev[i] += tmp[-1]
+        status[i] = -3 * (~np.isfinite(S[i]))
+
+    if np.any(ni):
+        args_indirect = [arg[ni] for arg in args]
+        tmp = _integral_bound(f, a[ni], b[ni], step[ni], args_indirect, constants)
+        S[ni], E[ni], status[ni] = tmp[:-1]
+        nfev[ni] += tmp[-1]
+
+    S, E = S.reshape(shape)[()], E.reshape(shape)[()]
+    status, nfev = status.reshape(shape)[()], nfev.reshape(shape)[()]
+
+    return OptimizeResult(sum=S, error=E, status=status, success=status == 0,
+                          nfev=nfev)
+
+
+def _direct(f, a, b, step, args, constants, inclusive=True):
+    # Directly evaluate the sum. To allow computation in a single
+    # vectorized call, we find the maximum number of points (over all
+    # slices) at which the function needs to be evaluated. In each slice,
+    # the function will be evaluated at the same number of points, but
+    # excessive points (those beyond the right sum limit `b`) are replaced
+    # with NaN. The function evaluated at NaN is NaN, and NaNs are ignored
+    # in the sum.
+    # In some cases it may be faster to loop over slices than to vectorize
+    # like this. This is an optimization that can be added later.
+    dtype, log, eps, zero, log2, rtol, atol, maxterms = constants
+    steps = np.round((b - a) / step) + int(inclusive)
+    max_steps = int(np.max(steps))
+    a2, b2, step2 = a[:, np.newaxis], b[:, np.newaxis], step[:, np.newaxis]
+    args2 = [arg[:, np.newaxis] for arg in args]
+    ks = a2 + np.arange(max_steps, dtype=dtype) * step2
+    i_nan = ks > b2
+    ks[i_nan] = np.nan
+    fs = f(ks, *args2)
+    fs[i_nan] = zero
+    nfev = max_steps-i_nan.sum(axis=-1)
+    S = special.logsumexp(fs, axis=-1) if log else np.sum(fs, axis=-1)
+    E = np.real(S) + np.log(eps) if log else abs(eps * S)
+    return S, E, nfev
+
+
+def _integral_bound(f, a, b, step, args, constants):
+    # find the number of direct steps to take until error is sufficiently small
+    dtype, log, eps, zero, log2, rtol, atol, maxterms = constants
+    a2 = a[..., np.newaxis]
+    step2 = step[..., np.newaxis]
+    args2 = [arg[..., np.newaxis] for arg in args]
+    n_steps = np.round(np.logspace(0, np.log10(maxterms), 10, dtype=dtype))
+    ks = a2 + n_steps * step2
+    fks = f(ks, *args2)
+    if fks.size and np.all((fks < atol)[..., -1]):
+        # find the location of a term that is less than the tolerance
+        nt = np.argmax(fks < atol, axis=-1)
+        nt = np.max(nt)
+    else:
+        nt = len(n_steps) - 1
+    n_steps = n_steps[nt]
+
+    # use integration to estimate the remaining sum
+    k = a + n_steps * step
+    left, left_error, nfev = _direct(f, a, k, step, args, constants, inclusive=False)
+    k[~np.isfinite(left)] = np.nan  # if sum is not finite, no sense in continuing
+    fk = fks[..., nt]
+    right = _tanhsinh(f, k, b, args=args, atol=atol, rtol=rtol, log=log)
```

Comment:
`atol`/`rtol` can be max of their existing values or based on min of the error terms